File: .skills/openclaw-skills/skills/0xnyk/xint/lib/format.ts
/**
 * Format tweets for Telegram or markdown output.
 */

import type { Tweet, UrlEntity } from "./api";

// --- Color support ---

let _colorOverride: boolean | undefined;

export function shouldColor(): boolean {
  if (process.env.NO_COLOR !== undefined) return false;
  if (_colorOverride !== undefined) return _colorOverride;
  if (process.env.TERM === "dumb") return false;
  return process.stdout.isTTY ?? false;
}

export function setColorOverride(enabled: boolean): void {
  _colorOverride = enabled;
}

export function cyan(s: string): string { return shouldColor() ? `\x1b[36m${s}\x1b[0m` : s; }
export function yellow(s: string): string { return shouldColor() ? `\x1b[33m${s}\x1b[0m` : s; }
export function dim(s: string): string { return shouldColor() ? `\x1b[2m${s}\x1b[0m` : s; }
export function bold(s: string): string { return shouldColor() ? `\x1b[1m${s}\x1b[0m` : s; }
export function green(s: string): string { return shouldColor() ? `\x1b[32m${s}\x1b[0m` : s; }
export function red(s: string): string { return shouldColor() ? `\x1b[31m${s}\x1b[0m` : s; }

export function stripAnsi(s: string): string {
  return s.replace(/\x1b\[[0-9;]*m/g, "");
}

function compactNumber(n: number): string {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(1)}K`;
  return String(n);
}

function timeAgo(dateStr: string): string {
  const diff = Date.now() - new Date(dateStr).getTime();
  const mins = Math.floor(diff / 60_000);
  if (mins < 60) return `${mins}m`;
  const hours = Math.floor(mins / 60);
  if (hours < 24) return `${hours}h`;
  const days = Math.floor(hours / 24);
  return `${days}d`;
}

function getMetrics(t: Tweet): Tweet["metrics"] {
  return (
    t.metrics || {
      likes: 0,
      retweets: 0,
      replies: 0,
      quotes: 0,
      impressions: 0,
      bookmarks: 0,
    }
  );
}

function getUsername(t: Tweet): string {
  return t.username || t.author_id || "unknown";
}

function getTweetUrl(t: Tweet): string {
  return t.tweet_url || `https://x.com/${getUsername(t)}/status/${t.id}`;
}

/**
 * Format a single tweet for Telegram (monospace-friendly).
 */
export function formatTweetTelegram(t: Tweet, index?: number, opts?: { full?: boolean }): string {
  const prefix = index !== undefined ? `${index + 1}. ` : "";
  const metrics = getMetrics(t);
  const engagement = `${compactNumber(metrics.likes)}❤️ ${compactNumber(metrics.impressions)}👁`;
  const time = timeAgo(t.created_at);

  // Truncate text to 200 chars for summary view, full text for single tweet/thread
  const text = opts?.full || t.text.length <= 200 ? t.text : t.text.slice(0, 197) + "...";
  // Clean up t.co links from text
  const cleanText = text.replace(/https:\/\/t\.co\/\S+/g, "").trim();

  let out = `${prefix}${cyan(`@${getUsername(t)}`)} (${yellow(engagement)} · ${dim(time)})\n${cleanText}`;

  if ((t.urls || []).length > 0) {
    const u = t.urls[0];
    if (u.title) {
      out += `\n📰 "${u.title}"`;
      if (u.description) out += ` — ${u.description.slice(0, 120)}`;
    }
    out += `\n🔗 ${u.url}`;
  }
  out += `\n${getTweetUrl(t)}`;

  return out;
}

/**
 * Format a list of tweets for Telegram.
 */
export function formatResultsTelegram(
  tweets: Tweet[],
  opts: { query?: string; limit?: number } = {}
): string {
  const limit = opts.limit || 15;
  const shown = tweets.slice(0, limit);

  let out = "";
  if (opts.query) {
    out += `🔍 "${opts.query}" — ${tweets.length} results\n\n`;
  }

  if (shown.length === 0) {
    out += "No results";
    return out;
  }

  out += shown.map((t, i) => formatTweetTelegram(t, i)).join("\n\n");

  if (tweets.length > limit) {
    out += `\n\n... +${tweets.length - limit} more`;
  }

  return out;
}

/**
 * Format a single tweet for markdown (research docs).
 */
export function formatTweetMarkdown(t: Tweet): string {
  const metrics = getMetrics(t);
  const username = getUsername(t);
  const tweetUrl = getTweetUrl(t);
  const cleanText = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim();
  const quoted = cleanText.replace(/\n/g, "\n  > ");
  const hashtags = (t.hashtags || []).map((h) => (h.startsWith("#") ? h : `#${h}`)).join(" ");

  let out = `- **@${username}** [Tweet](${tweetUrl})\n  > ${quoted}`;
  out += `\n  Metrics: ${metrics.likes} likes · ${metrics.retweets} retweets · ${metrics.replies} replies · ${metrics.impressions} impressions`;
  if (hashtags) out += `\n  Hashtags: ${hashtags}`;

  if ((t.urls || []).length > 0) {
    out += `\n  Links: ${(t.urls || []).map((u) => `[${u.title || new URL(u.url).hostname}](${u.url})`).join(", ")}`;
  }

  return out;
}

/**
 * Format results as a full markdown research document.
 */
export function formatResearchMarkdown(
  query: string,
  tweets: Tweet[],
  opts: {
    themes?: { title: string; tweetIds: string[] }[];
    apiCalls?: number;
    queries?: string[];
  } = {}
): string {
  const date = new Date().toISOString().split("T")[0];

  let out = `# X Research: ${query}\n\n`;
  out += `**Date:** ${date}\n`;
  out += `**Tweets found:** ${tweets.length}\n\n`;

  if (opts.themes && opts.themes.length > 0) {
    for (const theme of opts.themes) {
      out += `## ${theme.title}\n\n`;
      const themeTweets = theme.tweetIds
        .map((id) => tweets.find((t) => t.id === id))
        .filter(Boolean) as Tweet[];
      out += themeTweets.map(formatTweetMarkdown).join("\n\n");
      out += "\n\n";
    }
  } else {
    // No themes — just list by engagement
    out += `## Top Results (by engagement)\n\n`;
    out += tweets
      .slice(0, 30)
      .map(formatTweetMarkdown)
      .join("\n\n");
    out += "\n\n";
  }

  out += `---\n\n## Research Metadata\n`;
  out += `- **Query:** ${query}\n`;
  out += `- **Date:** ${date}\n`;
  if (opts.apiCalls) out += `- **API calls:** ${opts.apiCalls}\n`;
  out += `- **Tweets scanned:** ${tweets.length}\n`;
  out += `- **Est. cost:** ~$${((tweets.length * 0.005)).toFixed(2)}\n`;
  if (opts.queries) {
    out += `- **Search queries:**\n`;
    for (const q of opts.queries) {
      out += `  - \`${q}\`\n`;
    }
  }

  return out;
}

/**
 * Format a user profile for Telegram.
 */
export function formatProfileTelegram(user: any, tweets: Tweet[]): string {
  const m = user.public_metrics || {};
  let badges = "";
  if (user.subscription_type) {
    const badge = user.subscription_type === "Premium+" ? "Premium+" : user.subscription_type === "Premium" ? "Premium" : user.subscription_type;
    badges += ` [${badge}]`;
  }
  let out = `👤 @${user.username} — ${user.name}${badges}\n`;
  out += `${compactNumber(m.followers_count || 0)} followers · ${compactNumber(m.tweet_count || 0)} tweets\n`;
  if (user.connection_status && user.connection_status.length > 0) {
    out += `Connection: ${user.connection_status.join(", ")}\n`;
  }
  if (user.description) {
    out += `${user.description.slice(0, 150)}\n`;
  }
  out += `\nRecent:\n\n`;
  out += tweets
    .slice(0, 10)
    .map((t, i) => formatTweetTelegram(t, i))
    .join("\n\n");

  return out;
}

// ---------------------------------------------------------------------------
// CSV output
// ---------------------------------------------------------------------------

function csvEscape(s: string): string {
  if (s.includes(",") || s.includes('"') || s.includes("\n")) {
    return '"' + s.replace(/"/g, '""') + '"';
  }
  return s;
}

/**
 * Format tweets as CSV with header row.
 */
export function formatCsv(tweets: Tweet[]): string {
  const header = "id,username,name,text,likes,retweets,replies,impressions,bookmarks,created_at,url,hashtags,mentions";
  const rows = tweets.map(t => {
    const cleanText = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim();
    return [
      t.id,
      csvEscape(t.username),
      csvEscape(t.name),
      csvEscape(cleanText),
      t.metrics.likes,
      t.metrics.retweets,
      t.metrics.replies,
      t.metrics.impressions,
      t.metrics.bookmarks,
      t.created_at,
      t.tweet_url,
      csvEscape(t.hashtags.join(";")),
      csvEscape(t.mentions.join(";")),
    ].join(",");
  });
  return [header, ...rows].join("\n");
}

/**
 * Backwards-compatible alias used by older tests/callers.
 */
export function formatResultsCsv(tweets: Tweet[]): string {
  return formatCsv(tweets);
}

// ---------------------------------------------------------------------------
// JSONL output (one JSON object per line)
// ---------------------------------------------------------------------------

/**
 * Format tweets as JSONL — one JSON object per line.
 * Optimized for piping into jq, awk, or other line-oriented tools.
 */
export function formatJsonl(tweets: Tweet[]): string {
  return tweets.map(t => JSON.stringify(t)).join("\n");
}

/**
 * Backwards-compatible JSON formatter.
 */
export function formatResultsJson(tweets: Tweet[]): string {
  return JSON.stringify(tweets, null, 2);
}

/**
 * Filter objects to only include specified dot-separated field paths.
 * Usage: filterFields(data, "id,text,metrics.likes")
 */
export function filterFields<T>(data: T, fields: string): T {
  if (!data || typeof data !== "object") return data;
  const fieldPaths = fields.split(",").map(f => f.trim());

  function pick(obj: Record<string, unknown>, paths: string[]): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const path of paths) {
      const parts = path.split(".");
      let current: unknown = obj;
      for (const part of parts) {
        if (current && typeof current === "object" && !Array.isArray(current)) {
          current = (current as Record<string, unknown>)[part];
        } else {
          current = undefined;
          break;
        }
      }
      if (current !== undefined) {
        let target = result;
        for (let i = 0; i < parts.length - 1; i++) {
          if (!target[parts[i]] || typeof target[parts[i]] !== "object") {
            target[parts[i]] = {};
          }
          target = target[parts[i]] as Record<string, unknown>;
        }
        target[parts[parts.length - 1]] = current;
      }
    }
    return result;
  }

  if (Array.isArray(data)) {
    return data.map(item =>
      item && typeof item === "object" ? pick(item as Record<string, unknown>, fieldPaths) : item
    ) as T;
  }
  return pick(data as Record<string, unknown>, fieldPaths) as T;
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/format.rs
use crate::models::Tweet;
use chrono::Utc;
use colored::Colorize;
use serde::Serialize;
use serde_json::{Map, Value};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn compact_number(n: u64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.1}K", n as f64 / 1_000.0)
    } else {
        n.to_string()
    }
}

fn time_ago(date_str: &str) -> String {
    let Ok(date) = chrono::DateTime::parse_from_rfc3339(date_str) else {
        return String::new();
    };
    let diff = Utc::now().signed_duration_since(date);
    let mins = diff.num_minutes();
    if mins < 60 {
        return format!("{mins}m");
    }
    let hours = diff.num_hours();
    if hours < 24 {
        return format!("{hours}h");
    }
    let days = diff.num_days();
    format!("{days}d")
}

fn clean_tco(text: &str) -> String {
    let re_like = |s: &str| -> String {
        let mut result = String::new();
        let mut rest = s;
        while let Some(pos) = rest.find("https://t.co/") {
            result.push_str(&rest[..pos]);
            rest = &rest[pos..];
            // Skip until whitespace or end
            match rest.find(|c: char| c.is_whitespace()) {
                Some(end) => rest = &rest[end..],
                None => {
                    rest = "";
                    break;
                }
            }
        }
        result.push_str(rest);
        result.trim().to_string()
    };
    re_like(text)
}

// ---------------------------------------------------------------------------
// Telegram / terminal format
// ---------------------------------------------------------------------------

pub fn format_tweet_terminal(t: &Tweet, index: Option<usize>, full: bool) -> String {
    let prefix = match index {
        Some(i) => format!("{}. ", i + 1),
        None => String::new(),
    };
    let engagement = format!(
        "{}\u{2764}\u{fe0f} {}\u{1f441}",
        compact_number(t.metrics.likes),
        compact_number(t.metrics.impressions)
    );
    let time = time_ago(&t.created_at);

    let text = if full || t.text.len() <= 200 {
        t.text.clone()
    } else {
        format!(
            "{}...",
            &t.text[..t
                .text
                .char_indices()
                .nth(197)
                .map(|(i, _)| i)
                .unwrap_or(t.text.len())]
        )
    };
    let clean_text = clean_tco(&text);

    let username_display = format!("@{}", t.username).cyan();
    let engagement_display = engagement.yellow();
    let time_display = time.dimmed();

    let mut out = format!(
        "{}{} ({} \u{00b7} {})\n{}",
        prefix, username_display, engagement_display, time_display, clean_text
    );

    if let Some(u) = t.urls.first() {
        if let Some(ref title) = u.title {
            out.push_str(&format!("\n\u{1f4f0} \"{title}\""));
            if let Some(ref desc) = u.description {
                let short = if desc.len() > 120 {
                    &desc[..120]
                } else {
                    desc.as_str()
                };
                out.push_str(&format!(" \u{2014} {short}"));
            }
        }
        out.push_str(&format!("\n\u{1f517} {}", u.url));
    }
    out.push_str(&format!("\n{}", t.tweet_url));
    out
}

pub fn format_results_terminal(tweets: &[Tweet], query: Option<&str>, limit: usize) -> String {
    let shown = &tweets[..tweets.len().min(limit)];
    let mut out = String::new();

    if let Some(q) = query {
        out.push_str(&format!(
            "\u{1f50d} \"{}\" \u{2014} {} results\n\n",
            q,
            tweets.len().to_string().yellow()
        ));
    }

    for (i, t) in shown.iter().enumerate() {
        if i > 0 {
            out.push_str("\n\n");
        }
        out.push_str(&format_tweet_terminal(t, Some(i), false));
    }

    if tweets.len() > limit {
        out.push_str(&format!("\n\n... +{} more", tweets.len() - limit));
    }

    out
}

// ---------------------------------------------------------------------------
// Profile
// ---------------------------------------------------------------------------

pub fn format_profile_terminal(user: &serde_json::Value, tweets: &[Tweet]) -> String {
    let username = user.get("username").and_then(|v| v.as_str()).unwrap_or("?");
    let name = user.get("name").and_then(|v| v.as_str()).unwrap_or("?");
    let desc = user
        .get("description")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let pm = user.get("public_metrics");
    let followers = pm
        .and_then(|m| m.get("followers_count"))
        .and_then(|v| v.as_u64())
        .unwrap_or(0);
    let tweet_count = pm
        .and_then(|m| m.get("tweet_count"))
        .and_then(|v| v.as_u64())
        .unwrap_or(0);

    let username_display = format!("@{username}").cyan();
    let followers_display = compact_number(followers).yellow();
    let tweet_count_display = compact_number(tweet_count).yellow();

    let subscription_badge = user
        .get("subscription_type")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| format!(" [{}]", s))
        .unwrap_or_default();

    let connection_info = user
        .get("connection_status")
        .and_then(|v| v.as_array())
        .filter(|arr| !arr.is_empty())
        .map(|arr| {
            let statuses: Vec<&str> = arr.iter().filter_map(|v| v.as_str()).collect();
            format!(" ({})", statuses.join(", "))
        })
        .unwrap_or_default();

    let mut out = format!(
        "\u{1f464} {} \u{2014} {}{}{}\n{} followers \u{00b7} {} tweets\n",
        username_display,
        name,
        subscription_badge,
        connection_info,
        followers_display,
        tweet_count_display
    );

    if !desc.is_empty() {
        let short_desc = if desc.len() > 150 { &desc[..150] } else { desc };
        out.push_str(short_desc);
        out.push('\n');
    }

    out.push_str("\nRecent:\n\n");

    for (i, t) in tweets.iter().take(10).enumerate() {
        if i > 0 {
            out.push_str("\n\n");
        }
        out.push_str(&format_tweet_terminal(t, Some(i), false));
    }

    out
}

// ---------------------------------------------------------------------------
// Markdown
// ---------------------------------------------------------------------------

pub fn format_tweet_markdown(t: &Tweet) -> String {
    let engagement = format!("{}L {}I", t.metrics.likes, t.metrics.impressions);
    let clean_text = clean_tco(&t.text);
    let quoted = clean_text.replace('\n', "\n  > ");

    let mut out = format!(
        "- **@{}** ({}) [Tweet]({})\n  > {}",
        t.username, engagement, t.tweet_url, quoted
    );

    if !t.urls.is_empty() {
        let links: Vec<String> = t
            .urls
            .iter()
            .map(|u| {
                let display = match &u.title {
                    Some(t) => t.clone(),
                    None => url::Url::parse(&u.url)
                        .ok()
                        .and_then(|p| p.host_str().map(String::from))
                        .unwrap_or_else(|| "link".to_string()),
                };
                format!("[{}]({})", display, u.url)
            })
            .collect();
        out.push_str(&format!("\n  Links: {}", links.join(", ")));
    }

    out
}

pub fn format_research_markdown(query: &str, tweets: &[Tweet], queries: &[&str]) -> String {
    let date = Utc::now().format("%Y-%m-%d").to_string();
    let mut out = format!("# X Research: {query}\n\n");
    out.push_str(&format!("**Date:** {date}\n"));
    out.push_str(&format!("**Tweets found:** {}\n\n", tweets.len()));

    out.push_str("## Top Results (by engagement)\n\n");
    for t in tweets.iter().take(30) {
        out.push_str(&format_tweet_markdown(t));
        out.push_str("\n\n");
    }

    out.push_str("---\n\n## Research Metadata\n");
    out.push_str(&format!("- **Query:** {query}\n"));
    out.push_str(&format!("- **Date:** {date}\n"));
    out.push_str(&format!("- **Tweets scanned:** {}\n", tweets.len()));
    out.push_str(&format!(
        "- **Est. cost:** ~${:.2}\n",
        tweets.len() as f64 * 0.005
    ));
    if !queries.is_empty() {
        out.push_str("- **Search queries:**\n");
        for q in queries {
            out.push_str(&format!("  - `{q}`\n"));
        }
    }

    out
}

// ---------------------------------------------------------------------------
// CSV
// ---------------------------------------------------------------------------

fn csv_escape(s: &str) -> String {
    if s.contains(',') || s.contains('"') || s.contains('\n') {
        format!("\"{}\"", s.replace('"', "\"\""))
    } else {
        s.to_string()
    }
}

pub fn format_csv(tweets: &[Tweet]) -> String {
    let header = "id,username,name,text,likes,retweets,replies,impressions,bookmarks,created_at,url,hashtags,mentions";
    let rows: Vec<String> = tweets
        .iter()
        .map(|t| {
            let clean_text = clean_tco(&t.text);
            format!(
                "{},{},{},{},{},{},{},{},{},{},{},{},{}",
                t.id,
                csv_escape(&t.username),
                csv_escape(&t.name),
                csv_escape(&clean_text),
                t.metrics.likes,
                t.metrics.retweets,
                t.metrics.replies,
                t.metrics.impressions,
                t.metrics.bookmarks,
                t.created_at,
                t.tweet_url,
                csv_escape(&t.hashtags.join(";")),
                csv_escape(&t.mentions.join(";")),
            )
        })
        .collect();

    std::iter::once(header.to_string())
        .chain(rows)
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn active_fields() -> Option<String> {
    std::env::var("XINT_FIELDS")
        .ok()
        .map(|v| v.trim().to_string())
        .filter(|v| !v.is_empty())
}

fn parse_field_paths(fields: &str) -> Vec<Vec<String>> {
    fields
        .split(',')
        .map(|field| field.trim())
        .filter(|field| !field.is_empty())
        .map(|field| field.split('.').map(|p| p.trim().to_string()).collect())
        .filter(|parts: &Vec<String>| !parts.is_empty() && parts.iter().all(|p| !p.is_empty()))
        .collect()
}

fn extract_path(value: &Value, path: &[String]) -> Option<Value> {
    if path.is_empty() {
        return Some(value.clone());
    }
    match value {
        Value::Object(map) => {
            let head = &path[0];
            let next = map.get(head)?;
            extract_path(next, &path[1..])
        }
        _ => None,
    }
}

fn insert_path(target: &mut Map<String, Value>, path: &[String], value: Value) {
    if path.is_empty() {
        return;
    }
    if path.len() == 1 {
        target.insert(path[0].clone(), value);
        return;
    }

    let key = path[0].clone();
    let entry = target
        .entry(key)
        .or_insert_with(|| Value::Object(Map::new()));
    if !entry.is_object() {
        *entry = Value::Object(Map::new());
    }
    if let Value::Object(obj) = entry {
        insert_path(obj, &path[1..], value);
    }
}

fn filter_object(value: &Value, paths: &[Vec<String>]) -> Value {
    let mut out = Map::new();
    for path in paths {
        if let Some(found) = extract_path(value, path) {
            insert_path(&mut out, path, found);
        }
    }
    Value::Object(out)
}

pub fn filter_fields(value: &Value, fields: &str) -> Value {
    let paths = parse_field_paths(fields);
    if paths.is_empty() {
        return value.clone();
    }

    match value {
        Value::Object(_) => filter_object(value, &paths),
        Value::Array(items) => Value::Array(
            items
                .iter()
                .map(|item| {
                    if item.is_object() {
                        filter_object(item, &paths)
                    } else {
                        item.clone()
                    }
                })
                .collect(),
        ),
        _ => value.clone(),
    }
}

pub fn print_json_pretty_filtered<T: Serialize>(value: &T) -> anyhow::Result<()> {
    let mut json = serde_json::to_value(value)?;
    if let Some(fields) = active_fields() {
        json = filter_fields(&json, &fields);
    }
    println!("{}", serde_json::to_string_pretty(&json)?);
    Ok(())
}
